### PR TITLE
Set Account Balance

### PR DESCRIPTION
**Pull request by CoderZhi**
# Description
Via SubBalance and AddBalance to prevent mis-assignment of the balance value. Based on https://github.com/iotexproject/iotex-core/pull/3397 and https://github.com/iotexproject/iotex-core/pull/3400

Fixes # (issue)

## Type of change

Please delete options that are not relevant.
- [] Bug fix (non-breaking change which fixes an issue)
- [] New feature (non-breaking change which adds functionality)
- [] Code refactor or improvement
- [] Breaking change (fix or feature that would cause a new or changed behavior of existing functionality)
- [] This change requires a documentation update

# How Has This Been Tested?

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration
- [] Test A
- [] Test B

**Test Configuration**:

- Firmware version:
- Hardware:
- Toolchain:
- SDK:

# Checklist:
- [] My code follows the style guidelines of this project
- [] I have performed a self-review of my code
- [] I have commented my code, particularly in hard-to-understand areas
- [] I have made corresponding changes to the documentation
- [] My changes generate no new warnings
- [] I have added tests that prove my fix is effective or that my feature works
- [] New and existing unit tests pass locally with my changes
- [] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: e2etest/bigint_test.go [hunk 1]
```diff
@@ -141,7 +143,7 @@ func prepare(bc blockchain.Blockchain, sf factory.Factory, ap actpool.ActPool, e
 	r.NoError(err)
 	selp, err := action.Sign(elp, priKey)
 	r.NoError(err)
-	r.NoError(ap.Add(context.Background(), selp))
+	r.Error(ap.Add(context.Background(), selp))
```
Comment (dustinxie) [inline e2etest/bigint_test.go:146]: why it has error now?

File: state/account.go [hunk 2]
```diff
@@ -93,13 +95,32 @@ func (st *Account) Deserialize(buf []byte) error {
 	return nil
 }
 
+// HasSufficientBalance returns true if balance is larger than amount
```
Comment (dustinxie) [inline state/account.go:98]: larger than or equal

File: state/account.go [hunk 3]
```diff
@@ -93,13 +95,32 @@ func (st *Account) Deserialize(buf []byte) error {
...
+	}
+	return amount.Cmp(st.Balance) <= 0
+}
+
 // AddBalance adds balance for account state
-func (st *Account) AddBalance(amount *big.Int) {
-	st.Balance.Add(st.Balance, amount)
+func (st *Account) AddBalance(amount *big.Int) error {
+	if amount == nil || amount.Sign() < 0 {
+		return errors.Wrapf(ErrInvalidAmount, "amount %s shouldn't be negative", amount.String())
```
Comment (dustinxie) [inline state/account.go:109]: `amount.String()` panic if amount == nil, maybe just return `ErrInvalidAmount`

File: state/account.go [hunk 4]
```diff
@@ -93,13 +95,32 @@ func (st *Account) Deserialize(buf []byte) error {
...
+	}
+	if st.Balance != nil {
+		st.Balance = new(big.Int).Add(st.Balance, amount)
```
Comment (dustinxie) [inline state/account.go:112]: `st.Balance.Add(st.Balance, amount)`

File: state/account_test.go [hunk 5]
```diff
@@ -71,7 +71,7 @@ func TestClone(t *testing.T) {
 	account := ss.Clone()
 	require.Equal(big.NewInt(200), account.Balance)
 
-	account.AddBalance(big.NewInt(100))
+	require.NoError(account.AddBalance(big.NewInt(100)))
```
Comment (dustinxie) [inline state/account_test.go:74]: add test case that `Add(-1) Sub(-1)` would fail

File: action/protocol/account/protocol.go [hunk 6]
```diff
@@ -119,7 +119,9 @@ func createAccount(sm protocol.StateManager, encodedAddr string, init *big.Int)
 	case nil:
 		return errors.Errorf("failed to create account %s", encodedAddr)
 	case state.ErrStateNotExist:
-		account.Balance = init
+		if err := account.AddBalance(init); err != nil {
```
Comment (dustinxie) [inline action/protocol/account/protocol.go:122]: can we add a `SetBalance()` to make it clear?

File: action/protocol/account/transfer_test.go [hunk 7]
```diff
@@ -78,9 +78,9 @@ func TestProtocol_HandleTransfer(t *testing.T) {
 	alfa := identityset.Address(28)
 	bravo := identityset.Address(29)
 	charlie := identityset.Address(30)
-	require.NoError(accountutil.StoreAccount(sm, alfa, &state.Account{
-		Balance: big.NewInt(50005),
-	}))
+	acct1 := &state.Account{}
+	require.NoError(acct1.AddBalance(big.NewInt(50005)))
```
Comment (dustinxie) [inline action/protocol/account/transfer_test.go:82]: `SetBalance()`

File: action/protocol/account/util/util.go [hunk 8]
```diff
@@ -40,7 +40,9 @@ func LoadOrCreateAccount(sm protocol.StateManager, addr address.Address) (*state
 		return &account, nil
 	}
 	if errors.Cause(err) == state.ErrStateNotExist {
-		account.Balance = big.NewInt(0)
+		if err := account.AddBalance(big.NewInt(0)); err != nil {
```
Comment (dustinxie) [inline action/protocol/account/util/util.go:43]: `SetBalance()`

File: action/protocol/execution/evm/contract_test.go [hunk 9]
```diff
@@ -252,9 +252,8 @@ func TestSnapshot(t *testing.T) {
 	testfunc := func(enableAsync bool) {
 		sm, err := initMockStateManager(ctrl)
 		require.NoError(err)
-		s := &state.Account{
-			Balance: big.NewInt(5),
-		}
+		s := &state.Account{}
+		require.NoError(s.AddBalance(big.NewInt(5)))
```
Comment (dustinxie) [inline action/protocol/execution/evm/contract_test.go:256]: `SetBalance()`

File: action/protocol/execution/evm/evmstatedbadapter.go [hunk 10]
```diff
@@ -365,8 +369,10 @@ func (stateDB *StateDBAdapter) Suicide(evmAddr common.Address) bool {
 		return false
 	}
 	// clears the account balance
-	s.Balance = nil
-	s.Balance = big.NewInt(0)
+	if err := s.SubBalance(s.Balance); err != nil {
```
Comment (dustinxie) [inline action/protocol/execution/evm/evmstatedbadapter.go:372]: `SetBalance()`

File: action/protocol/staking/handlers_test.go [hunk 11]
```diff
@@ -2660,7 +2660,12 @@ func setupAccount(sm protocol.StateManager, addr address.Address, balance int64)
 	if err != nil {
 		return err
 	}
-	account.Balance = unit.ConvertIotxToRau(balance)
+	if err := account.SubBalance(account.Balance); err != nil {
+		return err
+	}
+	if err := account.AddBalance(unit.ConvertIotxToRau(balance)); err != nil {
+		return err
+	}
```
Comment (dustinxie) [inline action/protocol/staking/handlers_test.go:2668]: `SetBalance()`

File: action/protocol/account/protocol.go (hunk 6 above)
Comment (CoderZhi) [reply action/protocol/account/protocol.go:122]: SetBalance doesn't make sense. For new account, add balance is sufficient; for existing account, reset balance is not reasonable

File: e2etest/bigint_test.go (hunk 1 above)
Comment (CoderZhi) [reply e2etest/bigint_test.go:146]: because the amount cannot be negative now

File: state/account.go [hunk 12]
```diff
@@ -93,13 +95,32 @@ func (st *Account) Deserialize(buf []byte) error {
...
+	} else {
+		st.Balance = new(big.Int).Set(amount)
+	}
+	return nil
 }
 
 // SubBalance subtracts balance for account state
 func (st *Account) SubBalance(amount *big.Int) error {
+	if amount == nil || amount.Cmp(big.NewInt(0)) < 0 {
+		return errors.Wrapf(ErrInvalidAmount, "amount %s shouldn't be negative", amount.String())
```
Comment (Liuhaai) [inline state/account.go:122]: `(nil).String()`?

File: action/protocol/account/protocol.go (hunk 6 above)
Comment (Liuhaai) [reply action/protocol/account/protocol.go:122]: when will `ErrStateNotExist` be returned for a existing account?

**Review by dustinxie — APPROVED** [review]

File: state/account.go (hunk 12 above)
Comment (dustinxie) [reply state/account.go:122]: it will return `<nil>` for `nil`

**Review by Liuhaai — APPROVED** [review]

**Pull request closed without merging**